src/Adapter/FilesystemItems.php

```diff
@@ -0,0 +1,209 @@
+use FluentDOM\Utility\Iterators\MapIterator;
+use InvalidArgumentException;
+use Iterator;
+use IteratorAggregate;
+use Libero\ContentApiBundle\Exception\ItemNotFound;
+use Libero\ContentApiBundle\Exception\UnexpectedVersionNumber;
+use Libero\ContentApiBundle\Exception\VersionNotFound;
+use Libero\ContentApiBundle\Model\ItemId;
+use Libero\ContentApiBundle\Model\ItemListPage;
+use Libero\ContentApiBundle\Model\Items;
+use Libero\ContentApiBundle\Model\ItemVersion;
+use Libero\ContentApiBundle\Model\ItemVersionNumber;
+use LimitIterator;
+use RuntimeException;
+use SplFileInfo;
+use Symfony\Component\Filesystem\Filesystem;
+use Traversable;
+use function array_pop;
+use function array_values;
+use function count;
+use function fopen;
+use function is_dir;
+use function iterator_count;
+use function iterator_to_array;
+use function md5_file;
+use function rsort;
+
+final class FilesystemItems implements IteratorAggregate, Items
+{
+    private $filesystem;
+    private $path;
+
+    public function __construct(string $path, Filesystem $filesystem)
+    {
+        $this->path = $path;
+        $this->filesystem = $filesystem;
+    }
+
+    public function add(ItemVersion $item) : void
+    {
+        $id = $item->getId();
+        $version = $item->getVersion();
+
+        try {
+            $next = $this->get($id)->getVersion()->next();
+        } catch (ItemNotFound $e) {
+            $next = ItemVersionNumber::fromInt(1);
+        }
+
+        if ($version > $next) {
+            throw new UnexpectedVersionNumber($id, $version, $next);
+        }
+
+        $this->filesystem->dumpFile("{$this->path}/{$id}/{$version->toInt()}.xml", $item->getContent());
+    }
+
+    public function remove(ItemId $id, ?ItemVersionNumber $version) : void
+    {
+        if (!is_dir("{$this->path}/{$id}")) {
+            return;
+        }
+
+        if (null === $version) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        try {
+            $highest = $this->get($id)->getVersion();
+        } catch (ItemNotFound $e) {
+            return;
+        }
+
+        if ($version->toInt() < $highest->toInt()) {
+            throw new UnexpectedVersionNumber($id, $version, $highest);
+        }
+
+        if (1 === $version->toInt()) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        $this->filesystem->remove("{$this->path}/{$id}/{$version->toInt()}.xml");
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (!is_dir("{$this->path}/{$id}")) {
+            throw new ItemNotFound($id);
+        }
+
+        if (null === $version) {
+            $versions = iterator_to_array($this->getVersions($id));
+            rsort($versions);
+            $version = $versions[0];
+        }
+
+        if (!$content = @fopen($file = "{$this->path}/{$id}/{$version->toInt()}.xml", 'rb')) {
```

Comment:
can check with `file_exists()` to avoid the `@` suppression which may hide other errors; if the file is there but it can't be read (e.g. permissions, ownership, disk is full of dust) it should be 50x rather than `VersionNotFound`